lib/SILOptimizer/Mandatory/TFDeabstraction.cpp

```diff
@@ -1873,89 +1871,186 @@ void TFDeabstraction::checkAndCanonicalizeAttributes() {
-                   diag::tf_op_misuse, error)
-            .highlight(loc.getSourceRange());
-          isError = true;
-
-          // If we have more specific information about what went wrong, emit
-          // notes.
-          if (it->second.getKind() == SymbolicValue::Unknown)
-            it->second.emitUnknownDiagnosticNotes(opInfo->inst->getLoc());
-          break;
-        }
+  std::string opName = opInfo.opName.str();
+  SmallVector<SILValue, 4> inputs;
+  SmallVector<GraphOperationAttribute, 4> attributes;
 
-        inst->setOperand(i, newVal);
-      }
+  for (unsigned i = 0, e = opInfo.operandClasses.size(); i != e; ++i) {
+    auto operand = inst->getOperand(i);
+    auto operandTy = operand->getType();
+    auto operandClass = opInfo.operandClasses[i];
+
+    // Collect and validate input operands.
+    if (opInfo.isInput(i)) {
+      inputs.push_back(operand);
+
+      // Each input gets a marker mangled into the op name, because we need to
+      // be able to distinguish between normal inputs and elements of an input
+      // list.
 
-      // Don't emit a second error for this op if we already emitted one.
-      if (isError)
+      // If this is tfc.scalarToTensor, then the input must be a valid scalar.
+      if (opInfo.opName == "tfc.scalarToTensor") {
+        auto scalarType = operandTy.getSwiftRValueType();
+        if (convertSwiftTypeToTF(scalarType) == 0) {
+          diagnoseInvalid("scalarToTensor requires scalar value; unrecognized"
+                          " type '" + scalarType->getString() +
+                          "' is not allowed");
+          return;
+        }
+        opName += ",s";
         continue;
+      }
 
-      // Check to see if the usage of this op looks ok.  If not, reject it with
-      // an error and ignore it.
-      auto error = opInfo->checkAndDiagnoseOperands();
-      if (!error.empty()) {
-        // TODO: improve the diagnostic to talk about the parameter label in the
-        // user code, not the internal op attribute.  The bookkeeping for this
-        // isn't obvious though.
-        auto loc = getUserSourceLocation(inst);
-        diagnose(fn.getModule().getASTContext(), loc.getSourceLoc(),
-                 diag::tf_op_misuse, error)
-          .highlight(loc.getSourceRange());
+      // Tensor operand type's are ok.
+      if (isTensorFlowValue(operandTy)) {
+        opName += ",i";
         continue;
       }
 
-      // If the tensor operation uses array parameters or has scalar values that
-      // are passed through memory, promote them to being simple arguments to
-      // make all subsequent analyses and promotion of the tensor operations
-      // simpler.
-      opInfo->canonicalizeOperands(/*configuration*/ nullptr);
+      // FIXME: We need to figure out our model for InputLists with
+      // GraphOperation instructions.
+      // Model input lists as an "L" entry in the name, followed by one "e" for
```

Comment:
would it be useful to collect this comment, and the use of ",s" and ",i" into a comment block (and even add to SIL.rst at some point), to serve as the "spec" of the graph op inst? this'll be useful as we manipulate that inst in the subsequent partitioning and graph lowering passes.

Reply:
This logic will be moving into GraphOperationDecoder, that's the intent of the InputMarker enum that I've started building there in the subsequent patch.  I'll clean this up when I do implement input lists.